### PR TITLE
fix(email-security): improve SPF, DKIM, and DMARC record validation (fixes #152)

### LINKED ISSUE #152: Improvement :-  SPF/DKIM/DMARC record validation + minor cleanup in email_security_check
## Overview

The current `email_security_check` implementation performs basic SPF, DKIM, and DMARC checks, but there are several **correctness and edge-case issues** that can produce false positives, false negatives, or misleading security assessments.

The goal is to make the existing tool more accurate and reliable without requiring a major architectural rewrite.

Contributors are welcome to work on individual items below. **Partial pull requests are explicitly welcome** — you do not need to implement every fix in this issue.

---

### Current behavior

```python
if any("v=dkim1" in r.lower() or "p=" in r.lower() for r in records):
    return selector
```

A TXT record is treated as a valid DKIM record if it merely contains `v=dkim1` **or** `p=` anywhere in the string. This has two consequences that share the same root cause:

- **False positives on unrelated TXT records.** Any TXT record containing the substring `p=` (which is not an unusual substring) gets misclassified as DKIM.
- **Revoked keys counted as active.** A key revoked per RFC convention with an empty `p=` value —
  ```text
  v=DKIM1; p=
  ```
  — still matches `"v=dkim1" in r.lower()` and is reported as found, even though there is no active public key.

Because both symptoms come from the same permissive check, they should be fixed together in a single validator.

### Expected behavior

Add a tag-aware validator, e.g.:

```python
def _is_valid_dkim_record(record: str) -> bool:
    ...
```

that confirms:

* `v=DKIM1` is present and correctly formatted.
* `p=` is present.
* `p=` contains a non-empty public key.

The scanner should be able to distinguish:

* **Active DKIM key**
* **Revoked/empty DKIM key** (`p=` present but empty)
* **Malformed DKIM record**
* **No DKIM record**

A response shape such as:

```json
{
    "found_selectors": [],
    "revoked_selectors": ["selector1"]
}
```

(or an equivalent structure) should preserve this distinction. Exact schema is up to the contributor.

Keep the implementation focused on this tool's needs rather than a full DKIM RFC parser.

---

### Current behavior

```python
spf_records = [
    r for r in txt_records
    if r.lower().startswith("v=spf1")
]
record = spf_records[0]
```

If a domain publishes more than one SPF record, the extras are silently dropped — the tool reports whichever record DNS happens to return first.

### Expected behavior

If more than one SPF record is found:

* Detect the condition.
* Do not silently choose the first record.
* Mark the SPF configuration as invalid/ambiguous.
* Preserve all discovered records so callers can inspect them.
* Add a recommendation, e.g. *"Multiple SPF records found — SPF configuration is invalid. Consolidate the SPF mechanisms into a single SPF record."*

```json
{
    "found": true,
    "valid": false,
    "records": ["...", "..."]
}
```

---

### Current behavior

```python
if "-all" in record_lower:
    return "fail"
if "~all" in record_lower:
    return "softfail"
if "?all" in record_lower:
    return "neutral"
if "+all" in record_lower:
    return "pass"
return "unknown"
```

This is substring matching, not mechanism parsing, and it mishandles several real cases:

* **False matches inside unrelated tokens.** A record containing something like `-allow` or a hostname ending in `-all.example.com` would falsely trigger `"fail"`.
* **No `all` mechanism at all** (e.g. `v=spf1 include:_spf.google.com`) currently returns `"unknown"` — indistinguishable from a genuinely unparseable record.
* **Bare `all` with no qualifier** (e.g. `v=spf1 all`) is equivalent to `+all` per the SPF spec, but none of the four substring checks match it, so it also falls through to `"unknown"`. This is a distinct bug from the missing-mechanism case above, not just the same symptom.
* `+all` is currently detected correctly (worth confirming it stays correct after the rewrite, since it's a real risk indicator worth flagging).

### Expected behavior

1. Tokenize the SPF record into mechanisms/modifiers.
2. Locate the actual terminating `all` mechanism (not a substring match).
3. Determine its qualifier: `-`, `~`, `?`, `+`, or none (= `+`).
4. Return the correct policy for each of `-all`, `~all`, `?all`, `+all`, and bare `all`.
5. Explicitly represent a **missing** `all` mechanism (e.g. `all_mechanism: missing`) as distinct from an unparseable/unknown record.
6. Continue to flag `+all` / bare `all` with a security warning, since both are effectively "pass everything."

---

### Current behavior

```python
dmarc_records = [
    r for r in txt_records
    if r.lower().startswith("v=dmarc1")
]
record = dmarc_records[0]
```

Same pattern as the SPF issue above — additional DMARC records are silently ignored.

### Expected behavior

* Detect multiple DMARC records.
* Do not arbitrarily select the first one.
* Mark the DMARC configuration as invalid/ambiguous.
* Preserve the discovered records.
* Add a recommendation, e.g. *"Multiple DMARC records found — DMARC configuration is ambiguous. Publish a single DMARC policy record."*

Handle this consistently with the SPF multi-record fix (#2) for schema/style parity.

---

### 5a. DMARC `p=` value — optional / nice-to-have

```python
policy = tags.get("p", "none").lower()
```

Any value is accepted as-is, so `v=DMARC1; p=banana` results in `policy = "banana"`. Note this is **not a scoring bug**: the scoring logic only branches on `policy == "reject"` or `"quarantine"`; anything else — including a garbage value like `"banana"` — already falls through to the same default score as `p=none`. So a malformed policy is already treated as unprotected, which is the correct security outcome.

The only gap is that the *reported* `policy` field shows the raw invalid string instead of flagging it. This is a reporting-accuracy nicety, not a correctness fix — leave it for a contributor who wants it, but it's not required.

If addressed, validate `p=` against `none`, `quarantine`, `reject` and mark anything else invalid, e.g.:

```json
{
    "found": true,
    "valid": false,
    "policy": "invalid"
}
```

### 5b. Improve `rua` validation — required

```python
if "rua" not in tags:
```

This only checks tag presence. `v=DMARC1; p=reject; rua=` has the tag but no actual reporting address, and currently passes this check.

Check for a non-empty value instead:

```python
if not tags.get("rua"):
```

Distinguish between: `rua` missing entirely, `rua` present but empty, and `rua` containing a value. Full URI validation is out of scope — the goal is just to stop treating empty `rua=` as configured.

### 5c. Minor code cleanup — required

Two small cleanups in `email_security_check()` / `_discover_dynamic_selectors()`:

* **Remove the redundant `domain.lower()` call.** `email_security_check()` currently does:
  ```python
  domain = domain.strip().lower()
  domain = normalize_domain(domain)
  ```
  `normalize_domain()` already handles normalization, so the standalone `.lower()` before it is redundant — and for internationalized domains, lowercasing raw Unicode input before it's been normalized/punycode-encoded is the wrong order of operations. Drop the `.lower()` line and let `normalize_domain()` own casing.

* **Move `import datetime` to the top of the file.** It's currently imported inline inside `_discover_dynamic_selectors()`:
  ```python
  import datetime
  current_year = datetime.datetime.now().year
  ```
  Move this to the module-level imports at the top of the file with the other imports, per standard style.

---

# Acceptance Criteria

### SPF
* [ ] Multiple SPF records are detected and not silently reduced to the first record.
* [ ] The actual `all` mechanism is parsed via tokenization, not substring matching.
* [ ] `-all`, `~all`, `?all`, `+all`, and bare `all` are all handled correctly.
* [ ] A missing `all` mechanism is distinguishable from an unparseable/unknown record.
* [ ] False matches inside unrelated substrings (e.g. `-allow`) no longer occur.

### DKIM
* [ ] DKIM detection validates tags rather than matching arbitrary `p=` substrings.
* [ ] `v=DKIM1` is validated.
* [ ] Empty `p=` values are not treated as active keys.
* [ ] Revoked/empty keys are distinguishable from active keys and from "no record."
* [ ] Existing best-effort selector discovery behavior remains intact.
* [ ] The tool does not claim DKIM is definitively absent just because none of the guessed selectors matched.

### DMARC
* [ ] Multiple DMARC records are detected and not silently reduced to the first record.
* [ ] Empty `rua=` is not treated as a configured reporting destination.
* [ ] Existing valid DMARC behavior continues to work.
* [ ] *(Optional)* `p=` is validated against `none`, `quarantine`, `reject`; anything else is reported invalid. Not required — invalid values already score the same as `p=none`.

### Cleanup
* [ ] Redundant `domain.lower()` call removed from `email_security_check()`; normalization is left entirely to `normalize_domain()`.
* [ ] `import datetime` moved from inside `_discover_dynamic_selectors()` to the top-level imports.

---
## Notes

These enhancements are intentionally grouped into a single tracking issue for visibility.

They are **independent improvements**, and contributors are encouraged to implement them in **multiple pull requests** rather than attempting to complete everything at once.

Update test file for the changes you applied

Files to be touched :-  `tools/email_security_tool` , `tests/test_email_securuity.py`
Follow contribuitng.md guidelines

### PR DESCRIPTION
**Pull request by AnandkumarMall**
## Closes

Closes #152

## Summary

Improves SPF, DKIM, and DMARC record validation accuracy and reliability per RFC specifications in `tools/email_security_tool.py`, preventing false positives, false negatives, and ambiguous configurations.

## What changed

- **SPF Improvements**:
  - Tokenized the SPF record to accurately parse the terminating `all` mechanism (`-all`, `~all`, `?all`, `+all`, bare `all`), preventing false matches inside unrelated substrings (e.g. `-allow`).
  - Added explicit `"missing"` status when a valid SPF record lacks an `all` mechanism, distinct from unparseable records (`"unknown"`).
  - Detected multiple SPF records: marks configuration invalid (`valid: False`), preserves all records, and provides actionable remediation guidance per RFC 7208.
- **DKIM Improvements**:
  - Implemented tag-aware DKIM validator confirming `v=DKIM1`, `p=`, and public key presence.
  - Distinguishes active keys (`found_selectors`), revoked/empty keys (`revoked_selectors`), malformed records, and unrelated TXT records containing `p=`.
- **DMARC Improvements**:
  - Detected multiple DMARC records: marks configuration ambiguous/invalid (`valid: False`), preserves all records, and recommends publishing a single DMARC record.
  - Improved `rua` tag validation to treat empty `rua=` as unconfigured and deduct score accordingly.
  - Added validation for `p=` values (`none`, `quarantine`, `reject`), flagging invalid values as `policy: "invalid"`.
- **Code Cleanups**:
  - Moved `import datetime` to top-level module imports.
  - Removed redundant `domain.lower()` in `email_security_check` in favor of `normalize_domain()`.
- **Test Coverage**:
  - Expanded `tests/test_email_security.py` covering multiple SPF records, multiple DMARC records, active vs revoked DKIM keys, empty `rua=`, invalid DMARC policy, and SPF tokenization edge cases.

## Verification

- [x] Ran locally and confirmed it works as expected
- [x] Added/updated tests for the changed behavior
- [x] All tests pass (`uv run pytest tests/test_email_security.py -v` -> 12 passed; `uv run pytest` -> 552 passed)
- [x] No unrelated changes included
- [x] Checked `CONTRIBUTING.md` and applied all changes

### CODE REVIEW

File: tools/email_security_tool.py [hunk 1]
```diff
@@ -1,230 +1,350 @@
...
+    if v_val is not None:
+        if v_val.lower() != "dkim1":
+            return "malformed", tags
+        if "p" not in tags:
+            return "malformed", tags
+        if tags["p"]:
+            return "active", tags
+        return "revoked", tags
+
+    return "not_dkim", tags
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: RFC 6376 Section 3.6.1: The v tag is optional but recommended. If it is omitted, the version defaults to DKIM1. If a record lacks a v tag but contains a valid p tag, it is a valid active record. Your current logic flags these legitimate records as `not_dkim`.

File: tools/email_security_tool.py [hunk 2]
```diff
@@ -1,230 +1,350 @@
...
+    parts = [p.strip() for p in record.split(";") if p.strip()]
+    tags: Dict[str, str] = {}
+    for part in parts:
+        if "=" in part:
+            k, v = part.split("=", 1)
+            tags[k.strip().lower()] = v.strip()
+
+    v_val = tags.get("v")
+    if v_val is not None:
+        if v_val.lower() != "dkim1":
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: As the comment in line 80 say the parser is following RFC 6376. But RFC 6376 Section 3.2 : The version tag v is explicitly defined as case-sensitive and must be exactly lowercase `DKIM1`. An email signature containing `v=dkim1` or `v=DKIM1`  is invalid. Your function would mistakenly accept a lowercase or mixed-case version tag as valid, which breaks compliance.

File: tools/email_security_tool.py [hunk 3]
```diff
@@ -1,230 +1,350 @@
...
+    - 'revoked': v=DKIM1 with empty p= key
+    - 'malformed': v= present but not DKIM1, or missing p= tag
+    - 'not_dkim': unrelated TXT record
+    """
+    parts = [p.strip() for p in record.split(";") if p.strip()]
+    tags: Dict[str, str] = {}
+    for part in parts:
+        if "=" in part:
+            k, v = part.split("=", 1)
+            tags[k.strip().lower()] = v.strip()
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: RFC 6376 Section 3.2: Tag names are strictly defined as alpha-num sequences without internal or surrounding whitespace. A record formatted as `v = DKIM1` is structurally malformed because of the space before the equal sign. Stripping the key hides structural defects that should cause the record to fail validation. Also Lowercases tag names, so invalid `P=key` is treated as valid `p=key`; DKIM tag names are case-sensitive. It also silently overwrites duplicate tags, although duplicates invalidate the entire tag list.

File: tools/email_security_tool.py [hunk 4]
```diff
@@ -1,230 +1,350 @@
...
+    Parse a TXT record for DKIM tags per RFC 6376.
+    Returns (status, tags) where status is one of:
+    - 'active': v=DKIM1 with non-empty p= key
+    - 'revoked': v=DKIM1 with empty p= key
+    - 'malformed': v= present but not DKIM1, or missing p= tag
+    - 'not_dkim': unrelated TXT record
+    """
+    parts = [p.strip() for p in record.split(";") if p.strip()]
+    tags: Dict[str, str] = {}
+    for part in parts:
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py:114]: Accepts `p=key`; `v=DKIM1`, though `v=` must be the first tag when present. according to RFC 6376

File: tools/email_security_tool.py [hunk 5]
```diff
@@ -1,230 +1,350 @@
...
+                dynamic_selectors.update(["selector1", "msft", "microsoft"])
+    except Exception:
+        pass  # Fail gracefully if MX discovery hits a hitch
+
+    return list(dynamic_selectors)
+
+
+def _parse_dkim_record(record: str) -> Tuple[str, Dict[str, str]]:
+    """
+    Parse a TXT record for DKIM tags per RFC 6376.
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: Suggestion :- Add a link for reference for RFC 6376 .

File: tools/email_security_tool.py [hunk 6]
```diff
@@ -1,230 +1,350 @@
...
+    - 'neutral' for ?all
+    - 'pass' for +all or bare all
+    - 'missing' if record starts with v=spf1 but contains no 'all' mechanism
+    - 'unknown' for invalid/unparseable records
+    """
+    if not record.lower().startswith("v=spf1"):
+        return "unknown"
+
+    tokens = record.strip().split()
+    for token in reversed(tokens):
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: Your function loops through the tokens backward (`reversed(tokens)`) and returns immediately on the first all mechanism it finds. RFC 7208 Section 4.6.2: Evaluates mechanisms sequentially from left to right.

File: tools/email_security_tool.py [hunk 7]
```diff
@@ -1,230 +1,350 @@
...
+        if t_lower in ("-all", "~all", "?all", "+all", "all"):
+            if t_lower == "-all":
+                return "fail"
+            if t_lower == "~all":
+                return "softfail"
+            if t_lower == "?all":
+                return "neutral"
+            if t_lower in ("+all", "all"):
+                return "pass"
+    return "missing"
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py:214]: If a record doesn't end in all, your code returns "missing" . RFC 7208 Section 6.1: An SPF record does not require an all mechanism if it includes a redirect modifier (e.g., `v=spf1 redirect=_://google.com`). The target domain's all policy is inherited instead.

File: tools/email_security_tool.py (hunk 6 above)
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: Your function loops through the tokens backward (`reversed(tokens)`) and returns immediately on the first all mechanism it finds. RFC 7208 Section 4.6.2: Evaluates mechanisms sequentially from left to right.

File: tools/email_security_tool.py [hunk 8]
```diff
@@ -1,230 +1,350 @@
...
+    Returns:
+    - 'fail' for -all
+    - 'softfail' for ~all
+    - 'neutral' for ?all
+    - 'pass' for +all or bare all
+    - 'missing' if record starts with v=spf1 but contains no 'all' mechanism
+    - 'unknown' for invalid/unparseable records
+    """
+    if not record.lower().startswith("v=spf1"):
+        return "unknown"
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: RFC 7208 Section 4.1: The version identifier must be lowercase (`v=spf1`). A record beginning with `V=SPF1` is structurally invalid and must be completely ignored by the receiving server. Your function incorrectly accepts mixed-case versions. Also `startswith("v=spf1")` still accepts malformed values such as `v=spf1foo` or `v=spf1`; `-all` as SPF records. That collapses an unparseable record into missing/a policy result, contrary to the issue's requirement that malformed/unknown records remain distinct from a valid SPF record with no all mechanism. Tokenize first, require the first token to equal `v=spf1`, apply the same predicate when collecting SPF records, and add regression tests.

File: tools/email_security_tool.py [hunk 9]
```diff
@@ -1,230 +1,350 @@
...
+        return {"found": False, "valid": False, "record": None, "records": [], "policy": None, "score": 0}
+
+    if len(dmarc_records) > 1:
+        recommendations.append(
+            "Multiple DMARC records found — DMARC configuration is ambiguous. Publish a single DMARC policy record."
+        )
+        return {
+            "found": True,
+            "valid": False,
+            "record": dmarc_records[0],
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: Same issue as in line 156

File: tools/email_security_tool.py [hunk 10]
```diff
@@ -1,230 +1,350 @@
...
+        return {"found": False, "valid": False, "record": None, "records": [], "policy": None, "score": 0}
+
+    if len(spf_records) > 1:
+        recommendations.append(
+            "Multiple SPF records found — SPF configuration is invalid. Consolidate the SPF mechanisms into a single SPF record."
+        )
+        return {
+            "found": True,
+            "valid": False,
+            "record": spf_records[0],
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: When multiple SPF records are found, `record: spf_records[0]` still silently selects the first record. Since the requirement says not to choose one, Make it to return `record: None` and preserve the candidates only in `records` This would make the invalid/ambiguous state explicit.

File: tools/email_security_tool.py [hunk 11]
```diff
@@ -1,230 +1,350 @@
...
+    for part in record.split(";"):
+        part = part.strip()
+        if "=" in part:
+            k, v = part.split("=", 1)
+            tags[k.strip().lower()] = v.strip()
+
+    raw_policy = tags.get("p")
+    valid = True
+    if raw_policy is None:
+        policy = "none"
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py:325]: RFC 9989 Section 4.7 (Current Standard): Your fallback logic matches the standard. The spec dictates: "If this tag is not present in an otherwise syntactically valid DMARC Policy Record, then the record is treated as if it included 'p=none'" . But adding a recommendation will be helpful here instead of labeling it as perfectly valid
```suggestion
        policy = "none"
        recommendations.append(
        "DMARC record is missing the 'p=' tag. While the new RFC 9989 standard defaults "
        "this to 'none', legacy mail systems may ignore this record entirely."
    )
```

File: tools/email_security_tool.py [hunk 12]
```diff
@@ -1,230 +1,350 @@
...
+
+    def check_selector(selector: str) -> Tuple[str, str | None]:
+        records, _failed = _query_txt(f"{selector}._domainkey.{domain}")
+        for r in records:
+            status, _tags = _parse_dkim_record(r)
+            if status == "active":
+                return selector, "active"
+            elif status == "revoked":
+                return selector, "revoked"
+        return selector, None
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: `_parse_dkim_record()` can return malformed, but this branch only preserves active and revoked; malformed selectors fall through to `None` and become indistinguishable from selectors with no DKIM record. Issue #152 requires the scanner to distinguish active keys, revoked/empty keys, malformed records, and no DKIM record. Collect malformed selectors into a malformed_selectors field (or equivalent), return it in the result, and add assertions for the `existing v=DKIM1-without-p=` test case.

File: tools/email_security_tool.py [hunk 13]
```diff
@@ -1,230 +1,350 @@
...
+        spf_score = 20
+        recommendations.append("SPF uses softfail (~all) — consider a hard fail (-all) for stronger protection")
+    elif policy in ("neutral", "pass"):
+        spf_score = 10
+        if policy == "pass":
+            recommendations.append("SPF policy is 'pass' (+all) — provides little protection.")
+        else:
+            recommendations.append(f"SPF policy is '{policy}' — provides little protection.")
+    elif policy == "missing":
+        spf_score = 15
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: A record with a `"missing"` or completely undefined all mechanism defaults to an implicit neutral or undefined state in real-world mail servers. However, practically, a missing terminator is a major structural omission or syntax oversight by a domain administrator. A broken/incomplete record (15) should not receive a higher health score than a completed, fully syntactically valid record set to "neutral" (10).

File: tools/email_security_tool.py [hunk 14]
```diff
@@ -1,230 +1,350 @@
...
+            tags[k.strip().lower()] = v.strip()
+
+    v_val = tags.get("v")
+    if v_val is not None:
+        if v_val.lower() != "dkim1":
+            return "malformed", tags
+        if "p" not in tags:
+            return "malformed", tags
+        if tags["p"]:
+            return "active", tags
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py]: This only verifies that `p=` is non-empty, so arbitrary text such as `v=DKIM1; p=not-a-public-key` is classified as an active key. Issue #152 requires `p=` to contain a non-empty public key. Please validate the normalized value as base64 public-key data, classify invalid values as malformed, and add a regression test. A full cryptographic-key parser is unnecessary here.

**Review by gaoharimran29-glitch — CHANGES_REQUESTED** [review]
Thanks for the focused implementation and accompanying regression coverage. The PR addresses the central SPF, DKIM, DMARC, and cleanup goals in #152 well.  

But there are few issues in this code for which i requested changes please see inline comments below and fix them.

A few broader RFC inconsistencies also exist in pre-existing project code. Those are acknowledged but are not being treated as blockers for this PR unless they are introduced or directly affected here; we will track the broader standards-alignment work in a follow-up issue.

File: tools/email_security_tool.py [hunk 15]
```diff
@@ -75,51 +75,127 @@ def _discover_dynamic_selectors(domain: str) -> List[str]:
...
+
+    for part in parts:
+        if "=" not in part:
+            return "malformed", tags
+
+        raw_k, raw_v = part.split("=", 1)
+
+        # RFC 6376 §3.2: Tag names are strictly defined as alpha-num sequences
+        # without internal or surrounding whitespace. A space before '=' is invalid.
+        if not raw_k.isalnum():
```
Comment (gaoharimran29-glitch) [inline tools/email_security_tool.py:136]: `isalnum()` does not match the RFC 6376 tag-name rule. It rejects valid extension tags containing _ (for example, `x_foo=value`), causing an otherwise valid DKIM key to be marked malformed. It also accepts invalid names beginning with a digit and non-ASCII letters. Please validate tag names with an ASCII pattern: first character [A-Za-z], followed by zero or more [A-Za-z0-9_]; unknown valid tags should be ignored.

**Review by gaoharimran29-glitch — CHANGES_REQUESTED** [review]
Thanks for addressing the requested changes—I appreciate the effort. There is just one remaining item; please see the inline comment. Once that is resolved, this PR will be ready to merge.

**Comment by gaoharimran29-glitch** [comment]
Just to clarify what happened here: this PR was merged inadvertently while I was preparing to merge another approved implementation for the same issue. I’ve since reverted the merge so that we can proceed with the other implementation and keep a single solution for this issue in the codebase.

I want to emphasize that this decision is **not a reflection of the quality of your implementation or your capabilities as a developer**. Your PR was reviewed and was a valid approach to solving the issue. The fact that it won’t remain merged is purely due to the situation around having two independent implementations for the same issue, and the decision to proceed with the alternate one.

Apologies for the confusion caused by the accidental merge, and thank you for taking the time to investigate and contribute a solution. I really appreciate the effort that went into this PR.
